contracts/staking/stakeManager/StakeManager.sol

```diff
@@ -102,45 +124,200 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
+        dynasty = newDynasty;
+        WITHDRAWAL_DELAY = newDynasty;
+        auctionPeriod = newDynasty.div(4);
+        replacementCoolDown = currentEpoch.add(auctionPeriod);
+    }
+
+    // Housekeeping function. @todo remove later
+    function stopAuctions(uint256 forNCheckpoints) public onlyGovernance {
+        replacementCoolDown = currentEpoch.add(forNCheckpoints);
+    }
+
+    function updateProposerBonus(uint256 newProposerBonus) public onlyGovernance {
+        logger.logProposerBonusChange(newProposerBonus, proposerBonus);
+        require(newProposerBonus <= MAX_PROPOSER_BONUS, "too big");
+        proposerBonus = newProposerBonus;
+    }
+
+    function updateSignerUpdateLimit(uint256 _limit) public onlyGovernance {
+        signerUpdateLimit = _limit;
+    }
+
+    function updateMinAmounts(uint256 _minDeposit, uint256 _minHeimdallFee) public onlyGovernance {
+        minDeposit = _minDeposit;
+        minHeimdallFee = _minHeimdallFee;
+    }
+
+    function drainValidatorShares(
+        uint256 validatorId,
+        address tokenAddr,
+        address payable destination,
+        uint256 amount
+    ) external onlyGovernance {
+        address contractAddr = validators[validatorId].contractAddress;
+        require(contractAddr != address(0x0), "not validator");
+        IValidatorShare(contractAddr).drain(tokenAddr, destination, amount);
+    }
+
+    function drain(address destination, uint256 amount) external onlyGovernance {
+        _transferToken(destination, amount);
+    }
+
+    function reinitialize(
+        address _NFTContract,
+        address _stakingLogger,
+        address _validatorShareFactory,
+        address _auctionImplementation
+    ) external onlyGovernance {
+        require(isContract(_auctionImplementation), "auction impl incorrect");
+        auctionImplementation = _auctionImplementation;
+        NFTContract = StakingNFT(_NFTContract);
+        logger = StakingInfo(_stakingLogger);
+        validatorShareFactory = ValidatorShareFactory(_validatorShareFactory);
+    }
+
+    /**
+        Public Methods
+     */
+
+    function topUpForFee(address user, uint256 heimdallFee) public onlyWhenUnlocked {
+        _transferAndTopUp(user, msg.sender, heimdallFee, 0);
     }
 
     function claimFee(
         uint256 accumFeeAmount,
         uint256 index,
         bytes memory proof
     ) public {
-        //Ignoring other params becuase rewards distribution is on chain
+        //Ignoring other params because rewards' distribution is on chain
         require(
             keccak256(abi.encode(msg.sender, accumFeeAmount)).checkMembership(index, accountStateRoot, proof),
             "Wrong acc proof"
         );
         uint256 withdrawAmount = accumFeeAmount.sub(userFeeExit[msg.sender]);
         _claimFee(msg.sender, withdrawAmount);
         userFeeExit[msg.sender] = accumFeeAmount;
-        require(token.transfer(msg.sender, withdrawAmount));
-    }
-
-    function stake(
```

Comment:
No `stake` function required? We can use `stakeFor`?

Reply:
Correct.